<!DOCTYPE html>

<html>

<head>

<meta charset="utf-8" />
<meta name="generator" content="pandoc" />
<meta http-equiv="X-UA-Compatible" content="IE=EDGE" />

<meta name="viewport" content="width=device-width, initial-scale=1" />



<title>README</title>

<script>// Pandoc 2.9 adds attributes on both header and div. We remove the former (to
// be compatible with the behavior of Pandoc < 2.8).
document.addEventListener('DOMContentLoaded', function(e) {
  var hs = document.querySelectorAll("div.section[class*='level'] > :first-child");
  var i, h, a;
  for (i = 0; i < hs.length; i++) {
    h = hs[i];
    if (!/^h[1-6]$/i.test(h.tagName)) continue;  // it should be a header h1-h6
    a = h.attributes;
    while (a.length > 0) h.removeAttribute(a[0].name);
  }
});
</script>

<style type="text/css">
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
span.underline{text-decoration: underline;}
div.column{display: inline-block; vertical-align: top; width: 50%;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
</style>






<style type="text/css">

div.csl-bib-body { }
div.csl-entry {
clear: both;
}
.hanging div.csl-entry {
margin-left:2em;
text-indent:-2em;
}
div.csl-left-margin {
min-width:2em;
float:left;
}
div.csl-right-inline {
margin-left:2em;
padding-left:1em;
}
div.csl-indent {
margin-left: 2em;
}
</style>

<style type="text/css">body {
background-color: #fff;
margin: 1em auto;
max-width: 700px;
overflow: visible;
padding-left: 2em;
padding-right: 2em;
font-family: "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
font-size: 14px;
line-height: 1.35;
}
#TOC {
clear: both;
margin: 0 0 10px 10px;
padding: 4px;
width: 400px;
border: 1px solid #CCCCCC;
border-radius: 5px;
background-color: #f6f6f6;
font-size: 13px;
line-height: 1.3;
}
#TOC .toctitle {
font-weight: bold;
font-size: 15px;
margin-left: 5px;
}
#TOC ul {
padding-left: 40px;
margin-left: -1.5em;
margin-top: 5px;
margin-bottom: 5px;
}
#TOC ul ul {
margin-left: -2em;
}
#TOC li {
line-height: 16px;
}
table {
margin: 1em auto;
border-width: 1px;
border-color: #DDDDDD;
border-style: outset;
border-collapse: collapse;
}
table th {
border-width: 2px;
padding: 5px;
border-style: inset;
}
table td {
border-width: 1px;
border-style: inset;
line-height: 18px;
padding: 5px 5px;
}
table, table th, table td {
border-left-style: none;
border-right-style: none;
}
table thead, table tr.even {
background-color: #f7f7f7;
}
p {
margin: 0.5em 0;
}
blockquote {
background-color: #f6f6f6;
padding: 0.25em 0.75em;
}
hr {
border-style: solid;
border: none;
border-top: 1px solid #777;
margin: 28px 0;
}
dl {
margin-left: 0;
}
dl dd {
margin-bottom: 13px;
margin-left: 13px;
}
dl dt {
font-weight: bold;
}
ul {
margin-top: 0;
}
ul li {
list-style: circle outside;
}
ul ul {
margin-bottom: 0;
}
pre, code {
background-color: #f7f7f7;
border-radius: 3px;
color: #333;
white-space: pre-wrap; 
}
pre {
border-radius: 3px;
margin: 5px 0px 10px 0px;
padding: 10px;
}
pre:not([class]) {
background-color: #f7f7f7;
}
code {
font-family: Consolas, Monaco, 'Courier New', monospace;
font-size: 85%;
}
p > code, li > code {
padding: 2px 0px;
}
div.figure {
text-align: center;
}
img {
background-color: #FFFFFF;
padding: 2px;
border: 1px solid #DDDDDD;
border-radius: 3px;
border: 1px solid #CCCCCC;
margin: 0 5px;
}
h1 {
margin-top: 0;
font-size: 35px;
line-height: 40px;
}
h2 {
border-bottom: 4px solid #f7f7f7;
padding-top: 10px;
padding-bottom: 2px;
font-size: 145%;
}
h3 {
border-bottom: 2px solid #f7f7f7;
padding-top: 10px;
font-size: 120%;
}
h4 {
border-bottom: 1px solid #f7f7f7;
margin-left: 8px;
font-size: 105%;
}
h5, h6 {
border-bottom: 1px solid #ccc;
font-size: 105%;
}
a {
color: #0033dd;
text-decoration: none;
}
a:hover {
color: #6666ff; }
a:visited {
color: #800080; }
a:visited:hover {
color: #BB00BB; }
a[href^="http:"] {
text-decoration: underline; }
a[href^="https:"] {
text-decoration: underline; }

code > span.kw { color: #555; font-weight: bold; } 
code > span.dt { color: #902000; } 
code > span.dv { color: #40a070; } 
code > span.bn { color: #d14; } 
code > span.fl { color: #d14; } 
code > span.ch { color: #d14; } 
code > span.st { color: #d14; } 
code > span.co { color: #888888; font-style: italic; } 
code > span.ot { color: #007020; } 
code > span.al { color: #ff0000; font-weight: bold; } 
code > span.fu { color: #900; font-weight: bold; } 
code > span.er { color: #a61717; background-color: #e3d2d2; } 
</style>




</head>

<body>




<h1 class="title toc-ignore">README</h1>



<div id="gosorensen" class="section level1">
<h1>goSorensen</h1>
<p>This R package implements inferential methods to compare gene lists
(proving equivalence) in terms of their biological significance as
expressed in the Gene Ontology GO. The compared gene lists are
characterized by cross-tabulation frequency tables of enriched GO terms.
Dissimilarity between gene lists is evaluated using the Sorensen-Dice
dissimilarity. The fundamental guiding principle is that two gene lists
are taken as similar if they share a significant proportion of common
enriched GO terms.</p>
<p>This inferential method, is developed and explained in the paper
<em>An equivalence test between features lists, based on the
Sorensen-Dice index and the joint frequencies of GO term
enrichment</em>, available online in <a href="https://rdcu.be/cOISz" class="uri">https://rdcu.be/cOISz</a></p>
<div id="installation-instructions" class="section level2">
<h2>Installation Instructions</h2>
<p>goSorensen package has to be installed with a working R version
(&gt;=4.3.3). Installation could take a few minutes on a regular desktop
or laptop. Package can be installed from Bioconductor as follows:</p>
<pre><code>if (!requireNamespace(&quot;goSorensen&quot;, quietly = TRUE)) {
    BiocManager::install(&quot;goSorensen&quot;)
}
library(goSorensen)</code></pre>
<div id="installing-dependencies-of-genomic-annotation-packages-in-bioconductor" class="section level3">
<h3>Installing Dependencies of Genomic Annotation Packages in
Bioconductor</h3>
<p>Before using <code>goSorensen</code>, the user must have adequate
knowledge of the species they intend to focus their analysis. The
genomic annotation packages available in Bioconductor provide all the
essential information about these species. For instance, if the research
involves mice, the user must install and activate the annotation package
<code>org.Mm.eg.db</code> in the following manner:</p>
<pre><code>if (!requireNamespace(&quot;org.Mm.eg.db&quot;, quietly = TRUE)) {
    BiocManager::install(&quot;org.Mm.eg.db&quot;)
    }</code></pre>
<p>Other species may include:</p>
<ul>
<li><code>org.Hs.eg.db</code>: Genome wide annotation for Humans.</li>
<li><code>org.At.tair.db</code>: Genome wide annotation for
Arabidopsis</li>
<li><code>org.Ag.eg.db</code>: Genome wide annotation for Anopheles</li>
<li><code>org.Bt.eg.db</code>: Genome wide annotation for Bovine</li>
<li><code>org.Ce.eg.db</code>: Genome wide annotation for Worm</li>
<li><code>org.Cf.eg.db</code>: Genome wide annotation for Canine</li>
<li><code>org.Dm.eg.db</code>: Genome wide annotation for Fly</li>
<li><code>org.EcSakai.eg.db</code>: Genome wide annotation for E coli
strain Sakai</li>
<li><code>org.EcK12.eg.db</code>: Genome wide annotation for E coli
strain K12</li>
<li><code>org.Dr.eg.db</code>: Genome wide annotation for Zebrafish</li>
<li><code>org.Gg.eg.db</code>: Genome wide annotation for Chicken</li>
<li><code>org.Mm.eg.db</code>: Genome wide annotation for Mouse</li>
<li><code>org.Mmu.eg.db</code>: Genome wide annotation for Rhesus</li>
<li>etcâ€¦</li>
</ul>
<p>Due to the extensive research conducted on the human species and the
examples documented in goSorensen about this species, the installation
of the goSorensen package automatically includes the annotation package
<code>org.Hs.eg.db</code> as a dependency. Therefore, there is no need
to install it previously. The user must install the appropriate package
to use genomic annotation for other species.</p>
<p>The goSorensen functions require genome annotation via the
<code>orgPackg</code> argument to annotate genes in GO terms for
enrichment analysis. For example, the enrichment contingency table can
be computed using the <code>buildEnrichTable</code> function.</p>
<pre><code>buildEnrichTable(..., orgPackg = &quot;org.Mm.eg.db&quot;)</code></pre>
</div>
<div id="gene-identifiers" class="section level3">
<h3>Gene Identifiers</h3>
<p><code>goSorensen</code> functions, such as
<code>buidEnrichTable</code>, that entail identifying annotations in a
GO term, which are essential for determining enrichment, require the
user to provide, in the argument <code>geneUniverse</code>, a vector
containing the identifiers of the universe of genes (wide genome) of the
species being analysed. These gene identifiers can be readily obtained
from the above-mentioned genomic annotation packages through the
function <code>keys</code>. For instance, one can acquire the ENTREZ
identifiers for the universe of genes in mice as follows:</p>
<pre><code># Load the package:
library(org.Mm.eg.db)

# Obtain the IDs:
mouseEntrezIDs &lt;- keys(org.Mm.eg.db, keytype = &quot;ENTREZID&quot;)</code></pre>
<p>Or, for the human species, as follows:</p>
<pre><code># Load the package:
library(org.Hs.eg.db)

# Obtain the IDs:
humanEntrezIDs &lt;- keys(org.Hs.eg.db, keytype = &quot;ENTREZID&quot;)</code></pre>
<p>And in this way, for any other species</p>
<p>Then, one can use this object in the <code>geneUniverse</code>
argument in whatever goSorensen function (e.g.,
<code>buidEnrichTable</code>) is needed:</p>
<pre><code>buildEnrichTable(..., orgPackg = &quot;org.Mm.eg.db&quot;, geneUniverse = mouseEntrezIDs)</code></pre>
</div>
</div>
<div id="contribution-guidelines" class="section level2">
<h2>Contribution Guidelines</h2>
<p>Contributions are welcome, if you wish to contribute or give ideas to
improve the package, please you can contact with maintainer (Pablo
Flores) to the address <code>p_flores@espoch.edu.ec</code>, and we can
discuss your suggestion.</p>
</div>
<div id="references" class="section level2">
<h2>References</h2>
<div id="refs" class="references">
<div id="goSorensen">
<p>Flores, P., Salicru, M., Sanchez-Pla, A., &amp; Ocana, J. (2022). An
equivalence test between features lists, based on the Sorensen-Dice
index and the joint frequencies of GO term enrichment. BMC
bioinformatics, 23(1), 1-21.</p>
</div>
</div>
</div>
</div>



<!-- code folding -->


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
